lib/pop.md

```diff
@@ -0,0 +1,433 @@
+improvements herein included are *multiple inheritance* and *default values*.
+Conversely, POP illuminates the *essence* of these object systems, that was once
+the topic of many debates and publications, in the retrospectively superior
+framework of dynamic lazy pure functional programming.
+
+### Plan of this essay
+
+The first section below,
+[Mapping common concepts with terminology](#mapping-common-concepts-with-terminology),
+introduces the concepts that we build upon, with
+a precise terminology we use in the rest of this document.
+
+We then describe [multiple inheritance](#multiple-inheritance),
+how it works, and what makes it useful.
+
+In a section [POP compared to other object systems](#pop-compared-to-other-object-systems),
+we compare our design to other designs:
+what it has in common with other Nix extension mechanisms
+that makes it interesting compared to other systems outside Nix, and
+what it has in common with other systems outside Nix
+that makes it interesting compared to those Nix extension mechanisms.
+
+Finally, we have a section that provides
+[Informal Types for POP](#informal-types-for-pop),
+in a notional type system with subtyping and some notion of type indexing.
+This is obviously expressible with dependent types.
+A type system that can usefully express these types, yet is weaker than dependent types,
+and maybe also has decidable or principal typing, is left as an exercise to the reader.
+
+We conclude with a TODO list for future work on POP.
+
+### Some historical context
+
+Nixpkgs already has many ad-hoc variants of the same extension mechanism,
+as notably seen in [lib/fixed-points.nix](fixed-points.nix) and
+[lib/customisation.nix](customisation.nix) and initially introduced in 2015.
+Surprisingly, several users have noticed how this extension mechanism is
+essentially isomorphic to the object system of Jsonnet (2014), introduced a
+year earlier. The equivalent design was not the product of purposeful choice,
+but an accident that suggests that they independently discovered a same
+essential concept in the space of software design. Indeed, the Nix extension
+mechanisms and the Jsonnet object system are themselves independent reinventions
+in the same dynamic lazy pure functional framework, of
+the classless object systems first seen in Thinglab (1979) or T (1981),
+once made popular by Self (1987) and JavaScript (1995) and these days known as
+"prototype object systems".
+
+The developers of Nix's extension mechanisms don't look like they ever intended
+these mechanisms to be object systems at all; and though they are now aware of
+how the "equivalence" these mechanisms are to such object systems, they seem
+to mostly dismiss the fact as a mere curiosity, trivial, irrelevant, or even
+annoying—a distraction. Instead, I suggest it can be a inspiration, both ways:
+Nix can learn from past object system designs to improve its modularity.
```

Comment:
I like this rationale of unifying the ad-hoc extension mechanisms. I think it should be moved into an RFC.

Reply:
I agree a RFC would be warranted to mark pop as more than "experimental" and "for use only in gerbil and/or packages whose maintainers have explicitly endorsed that object system".

But I depend on pop heavily and would rather not have to either wait for a pop-only PR or strip it from my code. Actually, pop would probably harder to justify adding if no code uses it yet. So that would make it a catch-22.

Reply:
At the very least, anything involving changes to `lib/` should be moved to a separate PR.

Reply:
OK, I created https://github.com/NixOS/nixpkgs/pull/116275 for POP with the first five commits in this set.